File: runme/mcgen/local/datan/datsrc/auxzfn.c
#include "datsrc.h"

void LIBFUNC auxzfn_(double P_T *x0, double P_T *x1, double P_T *xzero,
             double (CBFUNC P_T *funct)(double const P_T *,
             double const P_T *, integer const P_T *, integer const P_T *),
             double const P_T *par, integer const P_T *npar1,
             integer const P_T *npar2, double const P_T *epsiln)
{
    /* System generated locals */
    double d__1;

    /* Local variables */
    integer i;
    double f0, f1, fm, xm, test;

    *xzero = *x0;
    for (i = 1; i <= 2000; ++i) {
        f0 = (*funct)(x0, par, npar1, npar2);
        f1 = (*funct)(x1, par, npar1, npar2);
        if (f0 == 0.) {
            *xzero = *x0;
            goto L20;
        } else if (f1 == 0.) {
            *xzero = *x1;
            goto L20;
        }
        xm = (*x0 + *x1) * .5;
        if ((d__1 = *x0 - *x1, abs(d__1)) >= *epsiln) {
            fm = (*funct)(&xm, par, npar1, npar2);
            test = f0 * fm;
            if (test < 0.) {
                *x1 = xm;
            } else {
                *x0 = xm;
            }
        } else {
            *xzero = xm;
            goto L20;
        }
    }
L20:
    return;
} /* auxzfn_ */

File: runme/mcgen/local/datan/datsrc/auxzfn.f
      SUBROUTINE AUXZFN(X0,X1,XZERO,FUNCT,PAR,NPAR1,NPAR2,EPSILN)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER (ZERO=0.D0, HALF=0.5D0)
      EXTERNAL FUNCT
      XZERO=X0
      DO 10 I=1,2000
        F0=FUNCT(X0,PAR,NPAR1,NPAR2)
        F1=FUNCT(X1,PAR,NPAR1,NPAR2)
        IF(F0.EQ.ZERO) THEN
          XZERO=X0
          GO TO 20
        ELSE IF(F1.EQ.ZERO) THEN
          XZERO=X1
          GO TO 20
        END IF
        XM=HALF*(X0+X1)
        IF(ABS(X0-X1).GE.EPSILN) THEN
          FM=FUNCT(XM,PAR,NPAR1,NPAR2)
          TEST=F0*FM
          IF(TEST .LT. ZERO) THEN
            X1=XM
          ELSE
            X0=XM
          END IF
        ELSE
          XZERO=XM
          GO TO 20
        END IF
   10 CONTINUE
   20 CONTINUE
      END
